### PR TITLE
Add ECS schema for user-agent ingest processor

### LINKED ISSUE #37329: Bring the user_agent processor in line with ECS
**Describe the context**

In developing ECS, we took a step back and took a critical look at the information we're looking for, when parsing a user agent. The current user_agent plugin for Ingest Node had a few issues, we found.

Let's start with the default parsing for Chrome 70.0.3538.102 on Mac 10.14.1:

`Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36`

Is parsed as such:

```JSON
{
  "patch" : "3538",
  "major" : "70",
  "minor" : "0",
  "os" : "Mac OS X 10.14.1",
  "os_minor" : "14",
  "os_major" : "10",
  "name" : "Chrome",
  "os_name" : "Mac OS X",
  "device" : "Other"
}
```

To follow ECS, the structure would ideally be:

```JSON
{
  "name" : "Chrome",
  "original" : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36",
  "os" : {
    "name" : "Mac OS X",
    "version" : "10.14.1",
    "full" : "Mac OS X 10.14.1"
  },
  "device" : "Other",
  "version" : "70.0.3538.102"
}
```

There's a few things to notice here

- By default, the versions are reported as a complete version string, not broken down. An option is still welcome, to output the breakdown of the version numbers.
  - Also: full version strings make support for pre-release versions trivial (e.g. `-beta1`, `-rc2` etc.)
- The original ua string is kept around, which can be made optional as well.

In translating multiple access log Filebeat modules to match ECS, I've had to repeatedly rename the fields around. Here are some examples:

* Apache - all fields renamed, but the version numbers are still broken down (which is not in ECS). Currently most access log modules are implemented this way.
  * [pipeline](https://github.com/elastic/beats/blob/529feda9377ca6df1da5c8f4bbaed89a7413158e/filebeat/module/apache2/access/ingest/default.json#L40-L80)
  * [result](https://github.com/elastic/beats/blob/529feda9377ca6df1da5c8f4bbaed89a7413158e/filebeat/module/apache2/access/test/test.log-expected.json#L36-L44)
* Traefik - fields renamed, and there's a hacky attempt at reconstructing full version fields :-)
  * [pipeline](https://github.com/elastic/beats/blob/529feda9377ca6df1da5c8f4bbaed89a7413158e/filebeat/module/traefik/access/ingest/pipeline.json#L71-L118)
  * [semi-successful version reconstruction](https://github.com/elastic/beats/blob/529feda9377ca6df1da5c8f4bbaed89a7413158e/filebeat/module/traefik/access/test/test.log-expected.json): Browsing around you'll see completely empty versions (`..`), partially reconstructed ones (`11.2.` instead of `11.2.5`) and successful reconstructions (`7.62.0`). Of course this can be cleaned up further to eliminate the noise.
  * Note that this has not been released yet, and will not be released in its current form, this is an experiment ;-)

I don't mind doing this in order to hit the 7.0 feature freeze in a pinch. But I think this will not be a good experience for users that try to follow ECS, when they use the user agent processor. They would benefit greatly from having the plugin follow ECS by default or via some easy to enable setting(s).

**Describe the feature**

Can we update the user agent parser to:

- output field names following the ECS schema
- output the full version string for the agent and for the OS
- output the original agent string at `.original`

I'm more than happy that we do this via one or more option flags.

cc @ruflin

### PR DESCRIPTION
**Pull request by dakrone**
This switches the format of the user agent processor to use the schema from [ECS](https://github.com/elastic/ecs).
So rather than something like this:

```
{
  "patch" : "3538",
  "major" : "70",
  "minor" : "0",
  "os" : "Mac OS X 10.14.1",
  "os_minor" : "14",
  "os_major" : "10",
  "name" : "Chrome",
  "os_name" : "Mac OS X",
  "device" : "Other"
}
```

The structure is now like this:

```
{
  "name" : "Chrome",
  "original" : "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36",
  "os" : {
    "name" : "Mac OS X",
    "version" : "10.14.1",
    "full" : "Mac OS X 10.14.1"
  },
  "device" : "Other",
  "version" : "70.0.3538.102"
}
```

This new can be configured by setting `"ecs": true` in the processor
configuration, and will be the default for 7.0. Leaving `ecs` unset or set as
`false` is deprecated.

Resolves #37329

This PR is against the 6.x branch and will be forward-ported with the deprecated parts removed (I will open a subsequent PR for that).

### CODE REVIEW

File: modules/ingest-user-agent/src/test/java/org/elasticsearch/ingest/useragent/UserAgentProcessorFactoryTests.java [hunk 1]
```diff
@@ -179,7 +185,7 @@ public void testInvalidProperty() throws Exception {
 
         ElasticsearchParseException e = expectThrows(ElasticsearchParseException.class, () -> factory.create(null, null, config));
         assertThat(e.getMessage(), equalTo("[properties] illegal property value [invalid]. valid values are [NAME, MAJOR, MINOR, "
-                + "PATCH, OS, OS_NAME, OS_MAJOR, OS_MINOR, DEVICE, BUILD]"));
+                + "PATCH, OS, OS_NAME, OS_MAJOR, OS_MINOR, DEVICE, BUILD, ORIGINAL, VERSION]"));
     }
```
Comment (webmat) [inline modules/ingest-user-agent/src/test/java/org/elasticsearch/ingest/useragent/UserAgentProcessorFactoryTests.java]: Is there also an `OS_VERSION` here?

File: modules/ingest-user-agent/src/test/resources/rest-api-spec/test/ingest-useragent/30_custom_regex.yml [hunk 2]
```diff
@@ -32,11 +33,6 @@
...
   - match: { _source.user_agent.name: "Test" }
-  - match: { _source.user_agent.os: "Other" }
-  - match: { _source.user_agent.os_name: "Other" }
   - match: { _source.user_agent.device: "Other" }
-  - is_false: _source.user_agent.os_major
-  - is_false: _source.user_agent.os_minor
-  - is_false: _source.user_agent.major
-  - is_false: _source.user_agent.minor
-  - is_false: _source.user_agent.patch
+  - is_false: _source.user_agent.os}
```
Comment (webmat) [inline modules/ingest-user-agent/src/test/resources/rest-api-spec/test/ingest-useragent/30_custom_regex.yml]: Extra closing brace?

**Review by webmat — COMMENTED** [review]
I'm not qualified to vet the ES code, but this looks great!

Thanks @dakrone. And once again, sorry for making the request so late in the process.

Noted two things below. Also, should you be introducing new options for `properties`, such as `original`?

File: modules/ingest-user-agent/src/test/java/org/elasticsearch/ingest/useragent/UserAgentProcessorFactoryTests.java (hunk 1 above)
Comment (dakrone) [reply modules/ingest-user-agent/src/test/java/org/elasticsearch/ingest/useragent/UserAgentProcessorFactoryTests.java]: For the `os` map, the `os` property now returns the entire map (sub fields aren't selectable)

**Comment by dakrone** [comment]
> Also, should you be introducing new options for properties, such as original?

These new properties are for the ECS format only, they're needed since the names have changed

**Review by webmat — APPROVED** [review]
LGTM

File: modules/ingest-user-agent/src/main/java/org/elasticsearch/ingest/useragent/UserAgentProcessor.java [hunk 3]
```diff
@@ -236,13 +310,31 @@ public UserAgentProcessor create(Map<String, Processor.Factory> factories, Strin
...
+            return new UserAgentProcessor(processorTag, field, targetField, parser, properties, ignoreMissing, useECS);
         }
     }
 
     enum Property {
 
-        NAME, MAJOR, MINOR, PATCH, OS, OS_NAME, OS_MAJOR, OS_MINOR, DEVICE, BUILD;
+        NAME,
+        // Deprecated in 6.7 (superceded by VERSION), to be removed in 7.0
+        @Deprecated MAJOR,
```
Comment (martijnvg) [inline modules/ingest-user-agent/src/main/java/org/elasticsearch/ingest/useragent/UserAgentProcessor.java]: should we also log deprecation warning if a deprecated property is used?

**Review by martijnvg — APPROVED** [review]
I left one question, but other than that this LGTM.

File: modules/ingest-user-agent/src/main/java/org/elasticsearch/ingest/useragent/UserAgentProcessor.java (hunk 3 above)
Comment (dakrone) [reply modules/ingest-user-agent/src/main/java/org/elasticsearch/ingest/useragent/UserAgentProcessor.java]: Sounds good, I added this, thanks for the reminder

**Comment by ruflin** [comment]
@dakrone Thanks for getting this in. Looking forward to have this in master so we can adjust our Filebeat pipelines.

**Pull request merged**